orttraining/orttraining/core/session/training_session.cc

```diff
@@ -880,6 +904,93 @@ common::Status TrainingSession::GetStateTensors(NameMLValMap& state_tensors) {
   return GetSessionState().GetInitializedTensors(GetStateTensorNames(), allow_missing, state_tensors);
 }
 
+common::Status TrainingSession::GetOptimizerState(std::unordered_map<std::string, NameMLValMap>& opt_state_tensors) {
+  const bool allow_missing = (opt_graph_config_.deepspeed_zero.stage != 0);
+  // weight_to_opt_mapping_ is in the format of {weight_name: {prefix: full_optimizer_name, ..}, ..}
+  for (const auto& weight_map : weight_to_opt_mapping_) {
+    std::unordered_set<std::string> opt_names;
+    for (const auto& opt_pair: weight_map.second) {
+      opt_names.emplace(opt_pair.second);
+    }
+    NameMLValMap curr_opt_tensors;
+    //std::unordered_map<std::string, std::string> opt_names(weight_map.second.begin(), weight_map.second.end());
+    const auto& weight_name = weight_map.first;
+    GetSessionState().GetInitializedTensors(opt_names, allow_missing, curr_opt_tensors);
+    opt_state_tensors[weight_name] = {};
+    // Keep only prefix in returned value
+    for (const auto& opt_pair: weight_map.second) {
+      const auto& opt_prefix = opt_pair.first;
+      const auto& opt_name = opt_pair.second;
+      //const std::string pre_fix_only = opt_name.substr(0, opt_name.find(weight_name) - 1);
+      opt_state_tensors[weight_name][opt_prefix] = curr_opt_tensors[opt_name];
+    }
+  }
+  return Status::OK();
+}
+
+common::Status TrainingSession::GetModelState(std::unordered_map<std::string, NameMLValMap>& model_state_tensors, bool include_mixed_precision_weights) {
+  const bool allow_missing = (opt_graph_config_.deepspeed_zero.stage != 0);
+  std::unordered_set<std::string> fp_tensor_names{};
+  fp_tensor_names.insert(
+      weights_to_train_.begin(), weights_to_train_.end());
+  // Add zero sharded weights, only needed for fp32 weights in mixed precision run
+  for (const auto& weight_sharded_pair: updated_weight_names_map_) {
+    fp_tensor_names.erase(weight_sharded_pair.first); // remove the original name
+    fp_tensor_names.insert(weight_sharded_pair.second);
+  }
+  NameMLValMap fp_weights;
+  GetSessionState().GetInitializedTensors(fp_tensor_names, allow_missing, fp_weights);
+  // Change key from sharded_name to weight_name
+  for (const auto& weight_sharded_pair: updated_weight_names_map_) {
+    const auto& it = fp_weights.find(weight_sharded_pair.second);
+    if (it != fp_weights.end()) {
+      fp_weights[weight_sharded_pair.first] = it->second;
+      fp_weights.erase(it);
+    }
+  }
+  model_state_tensors["full_precision"] = fp_weights;
```

Comment:
the frontend expects this key to be "fp32"

Reply:
Changed